### PR TITLE
Coupon Management: Networking - Coupon model

### PR DESCRIPTION
**Pull request by joshheald-a8c**
Part of #3912

## Description
Adds the Coupon model according to the API docs, except for a few fields which would add confusion and are unlikely to be required, e.g. the dates in the site’s timezone.

Synthesised `Codable` conformance is used as far as possible, which gives us more compiler-provided safety, e.g. around key decoding correctness. Overrides for keys are only given where they're strictly required, and they're in camel case rather than snake case to support using a decoder set to `.convertFromSnakeCase`.

`siteId` is a `private(set) var`, rather than a `let` as we would ideally like. This is required so that we can inject the `siteId` in the mapper, while still benefitting from the synthesised `decode` method. Injection in the mapper will be done via `.copy()`, not mutation, and future improvements in Swift may eventually let us make this a `let` again.

I'll put full details of this approach in a P2 and link to it from here, as it is a slight departure from the established convention. I think that doing it this way would make us less likely to make mistakes while writing model property names out in slightly different ways in several places through the file and tests.

## Testing
The model is not in use yet, so no applicable testing other than the unit tests, and CI once it's merged to the main repo.

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Model/Coupon.swift
Comment (pmusolino) [inline Networking/Networking/Model/Coupon.swift]: nit: also if it adds some extra lines, would you like to add a return after every property? I think it becomes more readable.

**Review by pmusolino — APPROVED** [review]
Congrats for opening your first PR on `WCiOS` @joshheald-a8c  🥳 

The code looks good to me :shipit: just a minor nit. I'll merge it after you will fix it ;) 

>  I thought I could spend the rest of today reviewing your PR from yesterday and catching up on some p2s and other things.

Since I don't want to block this PR, and since I'm merging it in a separate branch `feat/coupon-management`, we can discuss these aspects in the P2 post, and from what comes out, we can correct it later.

File: Networking/Networking/Model/Coupon.swift
Comment (joshheald-a8c) [reply Networking/Networking/Model/Coupon.swift]: Thanks! Yep, definitely will do that too, will be a little easier to do it after the other networking PRs are merged.

Comment (joshheald-a8c) [reply Networking/Networking/Model/Coupon.swift]: Done 😊

**Pull request merged**
